meetings/working-groups/roles/extension-wg-2024-06-07.md

```diff
@@ -0,0 +1,146 @@
+
+# Extensions WG 2024-06-07
+
+## Type erasure encoding
+
+We need to encode the extension type (for extension-aware compiler) and the underlying type (erasure and for non-extension-aware tools).
+
+We'll need to encode tuple names, dynamic bits, nullability, native integer for both types. 
+Ideally, this should be done in a way that leave any tools that understand current attribute/constructor.
+
+Options for encoding extension type:
+- generic attribute (pinging David to ask limitations/bugs on older frameworks)
+- attribute with typeof (may not refer to type parameters)
+- nested type
+- nested type with parallel method
+
+Decision: we'll spec out the parallel method in nested type design.
+
+### modopt
+```
+// source
+void M(E e)
+
+// codegen
+void M(C modopt(E) e)
+```
+
+```
+virtual void M2(C c)
+// other assembly
+override void M2(E e)
+```
+
+Con: binary break to change API from extension to underlying type and back
+Con: problem for override at language level (might be solvable)
+
+### Generic attribute
+```
+// source
+void M(E e)
+
+// codegen
+void M([Extension<E>] Underlying e) // System.Extension`1 (new)
+```
+Attribute doesn't exist downlevel (we should synthesize)
+Generic attribute not supported before .NET 7 (prove there's enough support, or block off)
+
+### Attribute with typeof
+
+Limitation with typeof in attribute referencing type parameter (typeof encoded as fully-qualified name, context-free, but no such syntax exists for type parameters)

```

Comment:
Won't this also lose the attributes that we use to encode nullability, dynamic, etc ... Basically I'm thinking of this as the following:

```csharp
// user writes
extension E<T> for List<T> { ... }
E<dynamic> M() => ...;

// compiler emits
[typeof(E<object>)]
```

Believe that essentially this approach also loses information encoded in attributes.